### DISCUSSION TITLE
Easy way to add niri to GDM on NixOS

### DISCUSSION THREAD

**Discussion by technologicalsingularity** [YaLTeR/niri#819] [Ideas]
Hey all ! First of all - niri is awesome! I've installed NixOS 24.05 stable from minimal ISO, then install GNOME by adding it in configuration.nix, then install 0.1.10-1 niri by root by `nix-env --install niri` from unstable nixpkgs. I'm new in nix, how to force niri to be shown in GDM menu ? (flakes enabled, without GUI nix build (by flake from master) and run ./result/.. has no effect - Permission denied ). I think some nix configuration. On debian 12  "niri --session" was be enough to start niri from pure terminal after boot in multi-user.target non graphical mode (systemctl set-default multi-user.target)

**Comment by YaLTeR** [comment]
@sodiboo any idea? Adding niri.desktop into `/usr/share/wayland-sessions/` should be enough to make it show up in GDM.

**Reply by sodiboo** [reply]
> Adding niri.desktop into `/usr/share/wayland-sessions/` should be enough to make it show up in GDM.

No such path exists on NixOS. However, the user states they did this:

> install 0.1.10-1 niri by root by nix-env --install niri from unstable nixpkgs

and that should be enough, indeed, because that package includes [the relevant file](https://github.com/NixOS/nixpkgs/blob/5a27cc051d111db0b33537d80cc2f8b8d2e687f2/pkgs/by-name/ni/niri/package.nix#L84) along with [magic metadata that I don't actually know what it does](https://github.com/NixOS/nixpkgs/blob/5a27cc051d111db0b33537d80cc2f8b8d2e687f2/pkgs/by-name/ni/niri/package.nix#L113)

**Comment by sodiboo** [comment]
Could you try installing `niri` by adding it to `environment.systemPackages` (or setting `programs.niri.enable = true;`) instead of using `nix-env`? I suspect that NixOS desktop sessions have some [additional metadata](https://github.com/NixOS/nixpkgs/blob/5a27cc051d111db0b33537d80cc2f8b8d2e687f2/pkgs/by-name/ni/niri/package.nix#L113) that needs to be passed at config time for a session to show up. I don't actually know for sure that this is the case, so it *might* not resolve your issue.

The `programs.niri.enable = true;` option also enables some other goodies that are nice to have, like a graphics driver and desktop portals.

Moreover, you probably shouldn't be using `nix-env`. Using an imperative package manager kinda defeats the purpose of NixOS. [Specifically `nix-env` is busted and breaks in subtle ways when nixpkgs changes its structure in most meaningful ways](https://stop-using-nix-env.privatevoid.net/). The primary purpose of this command (and its younger sibling `nix profile`, who is clearly the favorite child among the two) is to allow you to use Nix as just a package manager without any declarative config, on any distro you want. On NixOS, it doesn't actually serve an important purpose because everything you could ever want from it, your configuration can already do better. You can still use it of course (maybe not for niri if my hypothesis is correct, but it'll work fine in general for most apps), because this is just a recommendation and I can't force you to do anything, I'm not your mom.

**Reply by technologicalsingularity** [reply]
You are absolutely right that using `nix-env --install` is not `NixOS way` (this is like `apt install` when I used to on debian). Please tell me where write config ? I've installed NixOS 24.5 stable few day ago. Adding `programs.niri.enable = true;` into /etc/nixos/configuration.nix bring up error:
```nushell
 error: The option `programs.niri' does not exist. Definition values:
       - In `/etc/nixos/configuration.nix':
           {
             enable = true;
           }
```
Now I think how make correction to this situation: I've installed many packages by `nix-env --install`, in ideal may be I must `nix-env --uninstall` and write them one by one to config, Can I do that by few commands in a short manner without listing them one by one?

**Reply by sodiboo** [reply]
Ah, yeah, this doesn't exist on nixos 24.05; it was introduced in 24.11 which is not yet stable.

You're putting it in the right place, the option is just new.

If you want, you can use [my niri flake](https://github.com/sodiboo/niri-flake), which has a similar module to the one from nixpkgs, and should work fine on 24.05. You said you were using flakes? Add to your configuration's `flake.nix` an input of `niri.url = "github:sodiboo/niri-flake";` and then in your config add `imports = [ inputs.niri.nixosModules.niri ];`. Then the option `programs.niri.enable` should be available and work as expected. Don't have a `flake.nix` in your configuration? it should look something like [this](https://wiki.nixos.org/wiki/Flakes#Using_nix_flakes_with_NixOS)

Also, be cautious of mixing in particular niri across channels of NixOS. it must be built against the correct version of mesa, or you will get a black screen when starting a session. niri from `nixos-unstable` will not work on `nixos-24.05` and vice versa. To resolve this, you can switch your entire system to `nixos-unstable` if you want. You can also use my flake, which builds the packages against your system packageset (and you don't need to do anything to enable this; by default the modules install the latest niri release built against your system, and *not* the package from nixpkgs. you can override it if you want the ones from nixpkgs, though.)

You can just add niri to `environment.systemPackages` instead of adding my module or switching release channels, but beware of the mesa version mismatch that can occur if you try to use the niri from `nixos-unstable`. (and also beware of the [additional things](https://github.com/NixOS/nixpkgs/blob/c06dd23aad5eb9e4314b4ed2104cff49d7466ff5/nixos/modules/programs/wayland/niri.nix) that the modules do, which you'll have to do manually in this case)

> Now I think how make correction to this situation: I've installed many packages by nix-env --install, in ideal may be I must nix-env --uninstall and write them one by one to config, Can I do that by few commands in a short manner without listing them one by one?

In general, no, this cannot be automatically done in a correct manner. `nix-env` installs packages by *derivation name*, which is not what is used to reference a package in a declarative config, and in general isn't unique either. Use [`nix-env --query`](https://nix.dev/manual/nix/2.18/command-ref/nix-env/query) to list the installed packages ([discussion i found this from; asking your same question](https://discourse.nixos.org/t/moving-from-nix-env-to-configuration-nix/36955)), but then you'll more or less have to go through them one-by-one to make sure you uninstall the correct package and also add the corresponding package to your config. It shouldn't really take too long for a given package, because the derivation name is *usually* just the same as the attribute path in nixpkgs.

**Reply by sodiboo** [reply]
> I suspect that NixOS desktop sessions have some [additional metadata](https://github.com/NixOS/nixpkgs/blob/5a27cc051d111db0b33537d80cc2f8b8d2e687f2/pkgs/by-name/ni/niri/package.nix#L113) that needs to be passed at config time for a session to show up.

seems this is indeed true, that attribute is important for modules regarding display managers. [see usage here](https://github.com/NixOS/nixpkgs/blob/c06dd23aad5eb9e4314b4ed2104cff49d7466ff5/nixos/modules/services/display-managers/default.nix#L192)

so i think my hypothesis is right. `providedSessions` is important for display managers, and that can't be used when installed from `nix-env`

**Reply by technologicalsingularity** [reply]
Ok, thanks a lot, I want to use latest unstable OS, nix-env --query shows 32 packages, all installed fron unstable channel, not a lot. I've seen that 25.05 latest unstable, right ? Can I make 25.05 from 24.05 without disk format (repratitition) ? If I can have I old OS in boot menu like after "nixos-rebuild switch" command I'd seen when reconfiguration ? Where NixOS version full list can I see ? I can install only from minimal ISO on portable SSD (I do not know why) and then if niri will start I do not need GNOME.

**Reply by steinuil** [reply]
A "channel" is just a branch in the [nixpkgs](https://github.com/NixOS/nixpkgs/) repository. The channel `nix-env` uses is different from the one that's used in your flake.

You don't need to change the `system.stateVersion` or reinstall NixOS to update to a new release. If you're managing your system config from a `flake.nix` you can switch to the latest unstable version of nixpkgs by setting `nixpkgs.url` in the `flake.nix` inputs to `"github:NixOS/nixpkgs/nixos-unstable"`.

Every time you want to update your system you can run `nix flake update` to sync with the latest version of nixos-unstable and then `sudo nixos-rebuild switch --flake .#` to update.

Don't forget the `--flake .#` part, otherwise the version of nixpkgs will be set with channels, which shouldn't be used if you're using flakes!

If you want to use another release of NixOS you can see the list of branches from GitHub where it says "master" in the top left, under the repository name, and then you replace `nixos-unstable` in `inputs.nixpkgs.url` with whatever release you want.
